### PR TITLE
Specialize ComputationalResult for internal use?

### PR DESCRIPTION
**Pull request by greyscaled**
## Proposed Changes
- Use specializations of `ComputationalResult` containing specialized data that can optionally be sent to clients
  - Can use across internal services

### CODE REVIEW

File: src/main/java/ca/vapurrmaid/discretemathapplications/domain/computation/PrimeFactorizationResult.java [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * @author vapurrmaid
+ */
+public class PrimeFactorizationResult extends ComputationalResult {
+
+    private final Map<Integer, Integer> primeFactors;
```
Comment (vapurrmaid) [inline src/main/java/ca/vapurrmaid/discretemathapplications/domain/computation/PrimeFactorizationResult.java]: `Map<Integer, Integer>` is a tally of each prime.

Example, `2 2 5` would map to:

```
{ 2: 2, 5: 1 }
```

**Comment by vapurrmaid** [comment]
Maybe it's best to just build an unpack function?

Example:

```
class PrimeFactorizationResult extends ComputationalResult {
    public <?> getFactors() {
        // unpack getMessage()
    }
}

File: src/main/java/ca/vapurrmaid/discretemathapplications/domain/computation/ComputationalResult.java [hunk 2]
```diff
@@ -12,7 +12,7 @@
  */
 @Data
 @JsonInclude(Include.NON_NULL)
-public final class ComputationalResult {
+public class ComputationalResult {
```
Comment (vapurrmaid) [inline src/main/java/ca/vapurrmaid/discretemathapplications/domain/computation/ComputationalResult.java]: 👍

File: src/main/java/ca/vapurrmaid/discretemathapplications/domain/computation/ComputationalResult.java [hunk 3]
```diff
@@ -25,7 +25,7 @@
      * @param resultIsLogicallyTrue Indicates if the result of the computation
      * is logically consistent with what the operation asks.
      */
-    public ComputationalResult(final boolean resultIsLogicallyTrue) {
+    public ComputationalResult(boolean resultIsLogicallyTrue) {
         this.message = "";
```
Comment (vapurrmaid) [inline src/main/java/ca/vapurrmaid/discretemathapplications/domain/computation/ComputationalResult.java]: 👍

File: src/main/java/ca/vapurrmaid/discretemathapplications/domain/computation/ComputationalResult.java [hunk 4]
```diff
@@ -37,7 +37,7 @@ public ComputationalResult(final boolean resultIsLogicallyTrue) {
      * is logically consistent with what the operation asks.
      * @param message The computed result itself or a description
      */
-    public ComputationalResult(final boolean resultIsLogicallyTrue, final String message) {
+    public ComputationalResult(boolean resultIsLogicallyTrue, @NonNull final String message) {
         this.message = message;
```
Comment (vapurrmaid) [inline src/main/java/ca/vapurrmaid/discretemathapplications/domain/computation/ComputationalResult.java]: 👍

File: src/main/java/ca/vapurrmaid/discretemathapplications/domain/computation/PrimeFactorizationResult.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    /**
+     * Constructor.
+     *
+     * @param resultIsLogicallyTrue Indicates if the result of the computation
+     * is logically consistent with what the operation asks.
+     * @param message The computed result itself or a description
+     */
+    public PrimeFactorizationResult(boolean resultIsLogicallyTrue, @NonNull String message) {
+        super(resultIsLogicallyTrue, message);
+    }
```
Comment (vapurrmaid) [inline src/main/java/ca/vapurrmaid/discretemathapplications/domain/computation/PrimeFactorizationResult.java]: 👍

File: src/main/java/ca/vapurrmaid/discretemathapplications/domain/computation/PrimeFactorizationResult.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+    /**
+     * Returns a tally, in a <code>HashMap<Integer, Integer></code> of each
+     * prime factor
```
Comment (vapurrmaid) [inline src/main/java/ca/vapurrmaid/discretemathapplications/domain/computation/PrimeFactorizationResult.java]: Update grammar:

```
Returns a tally, in the form of a <code>HashMap<Integer, Integer></code>, of each
prime factor.
```

File: src/main/java/ca/vapurrmaid/discretemathapplications/domain/computation/PrimeFactorizationResult.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+     *
+     * @return tally of prime factors
+     */
+    public HashMap<Integer, Integer> getFactors() {
+        HashMap<Integer, Integer> factors = new HashMap<>();
+        String stringFactors = getMessage().split("=")[1].trim();
+        stringFactors = stringFactors.replace("\u22C5", "");
+        stringFactors = stringFactors.replace(" ", "");
```
Comment (vapurrmaid) [inline src/main/java/ca/vapurrmaid/discretemathapplications/domain/computation/PrimeFactorizationResult.java]:
```java
HashMap<Integer, Integer> factors = new HashMap<>();

// getMessage() is in the form:    number = factor \u22C5 factor ...
String stringFactors = getMessage().split("=")[1].trim();
stringFactors = stringFactors.replace("\u22C5", "");
stringFactors = stringFactors.replace(" ", "");
```

File: src/main/java/ca/vapurrmaid/discretemathapplications/domain/computation/PrimeFactorizationResult.java [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    /**
+     * Returns a tally, in a <code>HashMap<Integer, Integer></code> of each
+     * prime factor
+     *
+     * @return tally of prime factors
+     */
+    public HashMap<Integer, Integer> getFactors() {
+        HashMap<Integer, Integer> factors = new HashMap<>();
```
Comment (vapurrmaid) [inline src/main/java/ca/vapurrmaid/discretemathapplications/domain/computation/PrimeFactorizationResult.java]:
```
// holds a tally of each factor
```

File: src/main/java/ca/vapurrmaid/discretemathapplications/services/PrimeServiceImpl.java [hunk 9]
```diff
@@ -88,7 +89,7 @@ public Computation primeFactorsOfNaturalNumber(NaturalNumber n) {
         int original = n.getNumberAsInteger();
 
         if (n.getNumberAsInteger() == 1) {
-            computation.setResult(new ComputationalResult(true, "1"));
+            computation.setResult(new PrimeFactorizationResult(true, "1 = 1"));
```
Comment (vapurrmaid) [inline src/main/java/ca/vapurrmaid/discretemathapplications/services/PrimeServiceImpl.java]: 👍

File: src/main/java/ca/vapurrmaid/discretemathapplications/services/PrimeServiceImpl.java [hunk 10]
```diff
@@ -120,7 +121,7 @@ public Computation primeFactorsOfNaturalNumber(NaturalNumber n) {
         factorization += n.getNumberAsInteger();
         computation.appendComputationalStep(new ComputationalStep("remainder is prime, stop", ""));
 
-        computation.setResult(new ComputationalResult(true, String.format("%d = %s", original, factorization)));
+        computation.setResult(new PrimeFactorizationResult(true, String.format("%d = %s", original, factorization)));
```
Comment (vapurrmaid) [inline src/main/java/ca/vapurrmaid/discretemathapplications/services/PrimeServiceImpl.java]: 👍

File: src/test/java/ca/vapurrmaid/discretemathapplications/domain/Computation/PrimeFactorizationResultTest.java [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+                        entry(2, 1),
+                        entry(5, 1));
+
+        assertThat(twenty.getFactors())
+                .containsExactly(
+                        entry(2, 2),
+                        entry(5, 1));
+    }
+
+}
```
Comment (vapurrmaid) [inline src/test/java/ca/vapurrmaid/discretemathapplications/domain/Computation/PrimeFactorizationResultTest.java]: 👍

File: src/test/java/ca/vapurrmaid/discretemathapplications/services/PrimeServiceImplTest.java [hunk 12]
```diff
@@ -57,7 +57,7 @@ public void testIsNaturalNumberPrime() throws NaturalNumberException {
     @Test
     public void testPrimeFactorsOfNaturalNumber() throws NaturalNumberException {
         // test edge case - 1
-        assertThat(primeService.primeFactorsOfNaturalNumber(new NaturalNumber(1)).getResult().getMessage()).isEqualTo("1");
+        assertThat(primeService.primeFactorsOfNaturalNumber(new NaturalNumber(1)).getResult().getMessage()).isEqualTo("1 = 1");
```
Comment (vapurrmaid) [inline src/test/java/ca/vapurrmaid/discretemathapplications/services/PrimeServiceImplTest.java]: 👍 - good catch, was inconsistent with all other possibilities

**Comment by vapurrmaid** [comment]
I think a better approach for future is re-working base class into an Interface, and then specifying `@JsonIgnore` etc at the implementation level.

```java
public Interface ComputationalResult {
    String getMessage();
    boolean resultILogicallyConsistent();
}
```

```java
public PrimeFactorizationResult implements ComputationalResult { }
public DivisibilityResult implements ComputationalResult { }
...
```

This way we're always guaranteed `resultIsLogicallyConsistent` and `message` in the JSON deserialization, but we can add new properties as we see fit.

- Method return in Controller can just guarantee the Interface
- Internal services can use the implementation
